p2p/simulations/connect_test.go

```diff
@@ -0,0 +1,190 @@
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package simulations
+
+import (
+	"testing"
+
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+)
+
+func newTestNetwork(t *testing.T, nodeCount int) (*Network, []enode.ID) {
+	adapter := adapters.NewSimAdapter(adapters.Services{
+		"noopwoop": func(ctx *adapters.ServiceContext) (node.Service, error) {
+			return NewNoopService(nil), nil
+		},
+	})
+
+	// create network
+	network := NewNetwork(adapter, &NetworkConfig{
+		DefaultService: "noopwoop",
+	})
+
+	// create and start nodes
+	ids := make([]enode.ID, nodeCount)
+	for i := range ids {
+		conf := adapters.RandomNodeConfig()
+		node, err := network.NewNodeWithConfig(conf)
+		if err != nil {
+			t.Fatalf("error creating node: %s", err)
+		}
+		if err := network.Start(node.ID()); err != nil {
+			t.Fatalf("error starting node: %s", err)
+		}
+		ids[i] = node.ID()
+	}
+
+	if len(network.Conns) > 0 {
+		t.Fatal("no connections should exist after just adding nodes")
+	}
+
+	return network, ids
+}
+
+func TestConnectToPivotNode(t *testing.T) {
+	net, ids := newTestNetwork(t, 2)
+	defer net.Shutdown()
+
+	pivot := ids[0]
+	net.SetPivotNode(pivot)
+
+	other := ids[1]
+	err := net.ConnectToPivotNode(other)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if net.GetConn(pivot, other) == nil {
+		t.Error("pivot and the other node are not connected")
+	}
+}
+
+func TestConnectToLastNode(t *testing.T) {
+	net, ids := newTestNetwork(t, 10)
+	defer net.Shutdown()
+
+	first := ids[0]
+	if err := net.ConnectToLastNode(first); err != nil {
+		t.Fatal(err)
+	}
+
+	last := ids[len(ids)-1]
+	for i, id := range ids {
+		if id == first || id == last {
+			continue
+		}
+
+		if net.GetConn(first, id) != nil {
+			t.Errorf("connection must not exits with node(ind: %v, id: %v)", i, id)
+		}
+	}
+
+	if net.GetConn(first, last) == nil {
+		t.Error("first and last node must be connected")
+	}
+}
+
+func TestConnectToRandomNode(t *testing.T) {
+	net, ids := newTestNetwork(t, 10)
+	defer net.Shutdown()
+
+	err := net.ConnectToRandomNode(ids[0])
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var cc int
+	for i, a := range ids {
+		for _, b := range ids[i:] {
+			if net.GetConn(a, b) != nil {
+				cc++
+			}
+		}
+	}
+
+	if cc != 1 {
+		t.Errorf("expected one connection, got %v", cc)
+	}
+}
+
+func TestConnectNodesFull(t *testing.T) {
+	net, ids := newTestNetwork(t, 12)
+	defer net.Shutdown()
+
+	err := net.ConnectNodesFull(ids)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	VerifyFull(t, net, ids)
```

Comment:
because its used in different packages. read the code

Reply:
Yes, but do we _need_ it in different packages? ;)

Reply:
@janos wow. `export_test.go` seems an interesting solution to the problem.
I guess you mean this trick: https://medium.com/@robiplus/golang-trick-export-for-test-aa16cbd7b8cd


Example, in prod code: https://golang.org/src/net/http/export_test.go

Neat.

Reply:
@frncmx Yes, exactly that. :)

Reply:
@janos I played a little bit with Go and read the article again.

I was sure the `*_test.go` get ignored by the compiler when building the main binary.
I was not sure what will happen with `test.go`. Well, it's built into the main binary.

So, I would recommend the `test.go` files - there are more than one -  to be renamed `helper_test.go` or `helpers_test.go`.*Which one you prefer?*

Why not `export_test.go`? Well we are kinda not doing the same thing I think. We want to expose helpers and keep the in separate file. We are not exposing private - prod - functions.

---

There is this testing paradigm the you should only test public functionality, so you can avoid hurtful refactors in the future.

That can be achieved easily. For example in `main_test.go` you just write `package main_test` instead of `package main` and all go tools will still work for you.

The article tackles the scenario, when in the above case you cannot avoid using some private behavior, so you expose some private fields. But you should be still writing your test in the `main_test` package.


What do you think?

Reply:
I did not mean white/black box testing.

White box testing is just "the internal structure/design/implementation of the item being tested is known to the tester.". Thus we choose our test cases based on our implementation, where we think the code will potentially break. That does not include reaching under to hood and tweak the state.

Actually, I think there is a problem whit our current way of testing. As it introduces coupling between components (we have sometimes many things in the same package) and also introduces coupling between the tests and the code. The latter is almost always bad. (At Emarsys we used to have the ground rule "never test private", just use the class's public interface. That might make hard to setup a test case, but that was a good sign of violating the single responsibility principle. Actually, I only had to test a private method once. :) So life is definitely possible that way and refactoring was always(/mostly) sweet.)

Reply:
I agree, packages with more things that can be separated by responsibility should be split for easier understanding, usage, testing and refactoring. Designing package api is as important and designing communication apis.